### ISSUE TITLE
required_reviewers on PRs

### ISSUE THREAD

**Issue by cblecker** [kubernetes/test-infra#8422]
Proposal:
blunderbuss currently selects a certain number of reviewers from the reviewer pool. By design it doesn't select the same reviewers for every PR. However, there may be use cases (high sensitivity parts of the code base, such as API changes) where a person should be cc'd on every PR that touches that part of the code.

I'd like to suggest that the OWNERS file be extended with a field for required_reviewers. Github users/teams that are listed in that section should always be added to the review list, in addition to the normal rotating reviewers (not instead of).

/area prow

**Comment by cblecker** [comment]
cc @BenTheElder @cjwagner @matthyx @fejta

**Comment by BenTheElder** [comment]
Interesting. I've had one previous request for a more advanced version of this to detect changes to flags and CC a particular reviewer.

This sounds useful and pretty straightforward to implement. Should it also support regex &/or globs so we can require reviews for specific files?

**Comment by cblecker** [comment]
Yes, I was thinking so. I believe blunderbuss supports this already, but I think the approvers plugin doesn't.

**Comment by cjwagner** [comment]
The [`no_parent_owners` field](https://github.com/kubernetes/test-infra/blob/1d1789f127517ea6778ff6e8a0eed2f0a848f509/prow/repoowners/repoowners.go#L48). May be of interest. It doesn't exactly match the behavior or what you are suggesting, but it does facilitate 'locking down' a sub directory.

**Comment by fejta** [comment]
This duplicate the behavior of CODE_OWNERS, does that matter?

**Comment by cblecker** [comment]
@cjwagner Yeah that's not quite what I'm looking for. I don't want to short circuit either the reviewers or approvers that would normally be assigned -- this would be additive.

@fejta It *is* similar to the CODEOWNERS behaviour with two key differences:
- CODEOWNERS [must have write access](https://help.github.com/articles/about-codeowners/)
- CODEOWNERS aren't additive. It runs through the file top to bottom, and the last line that matches the file is chosen.

**Comment by fejta** [comment]
> CODEOWNERS must have write access

👎 🙄 oh github

**Comment by BenTheElder** [comment]
Wow, did not know about that write access requirement, that is a major point to mention next time someone asks about the difference...

**Comment by matthyx** [comment]
Do we have a place to document these features? I saw @cjwagner speaking about doc improvement efforts during last SIG.

**Comment by cjwagner** [comment]
We have some documentation about OWNERS files here: https://github.com/kubernetes/test-infra/tree/master/prow/plugins/approve/approvers
I think that OWNERs file docs (not approve plugin specific) should live here though: https://github.com/kubernetes/test-infra/tree/master/prow/repoowners

**Comment by cblecker** [comment]
I think we have consensus to move forward on this one. Marking as available for someone to pick up.

/help

**Comment by matthyx** [comment]
Ok, then :-D
/remove-help /assign

**Comment by cblecker** [comment]
You can do both commands in a single comment, they just need to be on separate lines

**Comment by spiffxp** [comment]
@matthyx 
> Do we have a place to document these features? I saw @cjwagner speaking about doc improvement efforts during last SIG

@cjwagner 
> We have some documentation about OWNERS files here: https://github.com/kubernetes/test-infra/tree/master/prow/plugins/approve/approvers
> I think that OWNERs file docs (not approve plugin specific) should live here though: https://github.com/kubernetes/test-infra/tree/master/prow/repoowners [ed: nothing here at the moment]

There is also https://github.com/kubernetes/community/blob/master/contributors/guide/owners.md

I would rather we have just one doc. I like the one I'm linking, but I'm biased since I wrote it (based on the approvers doc, I honestly just forgot to nuke that).  I recognize it mixes up the owners spec + how kubernetes uses owners + tools that use owners.  And it's a doc, so predictably it's fallen out of date (regex filters aren't mentioned at all).  Separate issue? I'm open to suggestion.
